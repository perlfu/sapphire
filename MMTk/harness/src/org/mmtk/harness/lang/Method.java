/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.mmtk.harness.lang;

/**
 * Abstract superclass of methods implemented in a variety of ways
 */
public abstract class Method implements Statement, Expression {

  /** The name of this block */
  protected final String name;
  /** Number of parameters */
  protected final int params;

  /**
   * Constructor.
   * @param name
   * @param params
   */
  protected Method(String name, int params) {
    this.name = name;
    this.params = params;
  }

  /**
   * Execute the statements in the method.
   */
  public void exec(Env env) throws ReturnException {
    exec(env, new Value[] {});
  }

  /**
   * {@link Expression#eval(Env)}
   * @param env
   * @return
   */
  public Value eval(Env env) {
    return eval(env, new Value[] {});
  }

  /**
   * Sub-class method to execute the method as a statement.
   * @param env
   * @param values
   */
  public abstract void exec(Env env, Value...values);

  /**
   * Sub-class method to evaluate the method as an expression.
   * @param env
   * @param values
   */
  public abstract Value eval(Env env, Value...values);

  /**
   * Get the name of this method.
   */
  public String getName() {
    return name;
  }

}
