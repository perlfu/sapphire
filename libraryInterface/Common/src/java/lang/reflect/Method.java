/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package java.lang.reflect;

import java.lang.annotation.Annotation;

import org.jikesrvm.classloader.*;
import org.jikesrvm.runtime.VM_Reflection;
import org.jikesrvm.runtime.VM_Magic;
import org.jikesrvm.runtime.VM_Runtime;

/**
 * Implementation of java.lang.reflect.Field for JikesRVM.
 *
 * By convention, order methods in the same order
 * as they appear in the method summary list of Sun's 1.4 Javadoc API.
 */
public final class Method extends AccessibleObject implements Member {
  final VM_Method method;

   // Prevent this class from being instantiated.
  private Method() {
    method = null;
  }

  // For use by JikesRVMSupport
  Method(VM_Method m) {
    method = m;
  }

  public boolean equals(Object other) {
    if (other instanceof Method) {
      return method == ((Method)other).method;
    } else {
      return false;
    }
  }

  public Class<?> getDeclaringClass() {
    return method.getDeclaringClass().getClassForType();
  }

  public Class<?>[] getExceptionTypes() {
    VM_TypeReference[] exceptionTypes = method.getExceptionTypes();
    if (exceptionTypes == null) {
      return new Class[0];
    } else {
      return JikesRVMSupport.typesToClasses(exceptionTypes);
    }
  }

  public int getModifiers() {
    return method.getModifiers();
  }

  public String getName() {
    return method.getName().toString();
  }

  public Class<?>[] getParameterTypes() {
    return JikesRVMSupport.typesToClasses(method.getParameterTypes());
  }

  public boolean isSynthetic() {
    return method.isSynthetic();
  }

  public Class<?> getReturnType() {
    return method.getReturnType().resolve().getClassForType();
  }

  public int hashCode() {
    int code1 = getName().hashCode();
    int code2 = method.getDeclaringClass().toString().hashCode();
    return code1 ^ code2;
  }

  public Object invoke(Object receiver, Object[] args) throws IllegalAccessException,
                                                              IllegalArgumentException,
                                                              ExceptionInInitializerError,
                                                              InvocationTargetException {
    VM_Method method = this.method;
    VM_Class declaringClass = method.getDeclaringClass();

    // validate "this" argument
    if (!method.isStatic()) {
      if (receiver == null) throw new NullPointerException();
      receiver = JikesRVMSupport.makeArgumentCompatible(declaringClass, receiver);
    }

    // validate number and types of remaining arguments
    VM_TypeReference[] parameterTypes = method.getParameterTypes();
    if (args == null) {
      if (parameterTypes.length != 0) {
        throw new IllegalArgumentException("argument count mismatch");
      }
    } else if (args.length != parameterTypes.length) {
      throw new IllegalArgumentException("argument count mismatch");
    }
    for (int i = 0, n = parameterTypes.length; i < n; ++i) {
      args[i] = JikesRVMSupport.makeArgumentCompatible(parameterTypes[i].resolve(), args[i]);
    }

    // Accessibility checks
    if (!method.isPublic() && !isAccessible()) {
      VM_Class accessingClass = VM_Class.getClassFromStackFrame(1);
      JikesRVMSupport.checkAccess(method, accessingClass);
    }

    // find the right method to call
    if (! method.isStatic()) {
        VM_Class C = VM_Magic.getObjectType(receiver).asClass();
        method = C.findVirtualMethod(method.getName(), method.getDescriptor());
    }

    // Forces initialization of declaring class
    if (method.isStatic() && !declaringClass.isInitialized()) {
      try {
        VM_Runtime.initializeClassForDynamicLink(declaringClass);
      } catch (Throwable e) {
        ExceptionInInitializerError ex = new ExceptionInInitializerError();
        ex.initCause(e);
        throw ex;
      }
    }

    // Invoke method
    try {
      return VM_Reflection.invoke(method, receiver, args);
    } catch (Throwable t) {
      throw new InvocationTargetException(t,
                "While invoking the method:\n\t\"" + method + "\"\n"
                + " on the object:\n\t\"" + receiver + "\"\n"
                + " it threw the exception:\n\t\"" + t + "\"");
     }
  }

  public String toString() {
    StringBuilder sb = new StringBuilder(128);
    Modifier.toString(getModifiers(), sb).append(' ');
    sb.append(JikesRVMHelpers.getUserName(getReturnType())).append(' ');
    sb.append(getDeclaringClass().getName()).append('.');
    sb.append(getName()).append('(');
    Class<?>[] c = getParameterTypes();
    if (c.length > 0)
      {
        sb.append(JikesRVMHelpers.getUserName(c[0]));
        for (int i = 1; i < c.length; i++)
          sb.append(',').append(JikesRVMHelpers.getUserName(c[i]));
      }
    sb.append(')');
    c = getExceptionTypes();
    if (c.length > 0)
      {
        sb.append(" throws ").append(c[0].getName());
        for (int i = 1; i < c.length; i++)
          sb.append(',').append(c[i].getName());
      }
    return sb.toString();
  }

  // AnnotatedElement interface

  public Annotation[] getDeclaredAnnotations() {
    return method.getDeclaredAnnotations();
  }

  public <T extends Annotation> T getAnnotation(Class<T> annotationClass) {
    return method.getAnnotation(annotationClass);
  }

  // Generics support

  public TypeVariable<?>[] getTypeParameters() {
    VM_Atom sig = method.getSignature();
    if (sig == null) {
      return new TypeVariable[0];
    } else {
      return JikesRVMHelpers.getTypeParameters(this, sig);
    }
  }

  public Type[] getGenericExceptionTypes() {
    VM_Atom sig = method.getSignature();
    if (sig == null) {
      return getExceptionTypes();
    } else {
      return JikesRVMHelpers.getGenericExceptionTypes(this, sig);
    }
  }

  public Type[] getGenericParameterTypes() {
    VM_Atom sig = method.getSignature();
    if (sig == null) {
      return getParameterTypes();
    } else {
      return JikesRVMHelpers.getGenericParameterTypes(this, sig);
    }
  }

  public Type getGenericReturnType() {
    VM_Atom sig = method.getSignature();
    if (sig == null) {
      return getReturnType();
    } else {
      return JikesRVMHelpers.getGenericReturnType(this, sig);
    }
  }

  public String toGenericString() {
    StringBuilder sb = new StringBuilder(128);
    Modifier.toString(getModifiers(), sb).append(' ');
    Constructor.addTypeParameters(sb, getTypeParameters());
    sb.append(getGenericReturnType()).append(' ');
    sb.append(getDeclaringClass().getName()).append('.');
    sb.append(getName()).append('(');
    Type[] types = getGenericParameterTypes();
    if (types.length > 0)
      {
        sb.append(types[0]);
        for (int i = 1; i < types.length; i++)
          sb.append(',').append(types[i]);
      }
    sb.append(')');
    types = getGenericExceptionTypes();
    if (types.length > 0)
      {
        sb.append(" throws ").append(types[0]);
        for (int i = 1; i < types.length; i++)
          sb.append(',').append(types[i]);
      }
    return sb.toString();
  }
}
