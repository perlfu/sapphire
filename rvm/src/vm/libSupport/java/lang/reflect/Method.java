/*
 * (C) Copyright IBM Corp 2002
 */
//$Id$
package java.lang.reflect;

import com.ibm.JikesRVM.classloader.*;
import com.ibm.JikesRVM.VM_Reflection;
import com.ibm.JikesRVM.VM_Magic;
import com.ibm.JikesRVM.VM_Runtime;

/**
 * Implementation of java.lang.reflect.Field for JikesRVM.
 *
 * By convention, order methods in the same order
 * as they appear in the method summary list of Sun's 1.4 Javadoc API. 
 *
 * @author John Barton 
 * @author Julian Dolby
 * @author Stephen Fink
 * @author Eugene Gluzberg
 * @author Dave Grove
 * @modified Steven Augart
 */
public final class Method extends AccessibleObject implements Member {
  final VM_Method method;

   // Prevent this class from being instantiated.
  private Method() {
    method = null;
  }
    
  // For use by JikesRVMSupport
  Method(VM_Method m) {
    method = m;
  }

  public boolean equals(Object other) { 
    if (other instanceof Method) {
      return method == ((Method)other).method;
    } else {
      return false;
    }
  }

  public Class getDeclaringClass() {
    return method.getDeclaringClass().getClassForType();
  }

  public Class[] getExceptionTypes() {
    VM_TypeReference[] exceptionTypes = method.getExceptionTypes();
    if (exceptionTypes == null) {
      return new Class[0];
    } else {
      return JikesRVMSupport.typesToClasses(exceptionTypes);
    }
  }

  public int getModifiers() {
    return method.getModifiers();
  }

  public String getName() {
    return method.getName().toString();
  }

  public Class[] getParameterTypes() {
    return JikesRVMSupport.typesToClasses(method.getParameterTypes());
  }

  public Class getReturnType() {
    return method.getReturnType().resolve().getClassForType();
  }

  public int hashCode() {
    int code1 = getName().hashCode();
    int code2 = method.getDeclaringClass().toString().hashCode();
    return code1 ^ code2;
  }

  public Object invoke(Object receiver, Object args[]) throws IllegalAccessException, 
                                                              IllegalArgumentException, 
                                                              ExceptionInInitializerError,
                                                              InvocationTargetException {
    VM_Method method = this.method;
    VM_Class declaringClass = method.getDeclaringClass();
    
    // validate "this" argument
    if (!method.isStatic()) {
      if (receiver == null) throw new NullPointerException();
      receiver = JikesRVMSupport.makeArgumentCompatible(declaringClass, receiver);
    }
    
    // validate number and types of remaining arguments
    VM_TypeReference[] parameterTypes = method.getParameterTypes();
    if (args == null) {
      if (parameterTypes.length != 0) {
        throw new IllegalArgumentException("argument count mismatch");
      }
    } else if (args.length != parameterTypes.length) {
      throw new IllegalArgumentException("argument count mismatch");
    }
    for (int i = 0, n = parameterTypes.length; i < n; ++i) {
      args[i] = JikesRVMSupport.makeArgumentCompatible(parameterTypes[i].resolve(), args[i]);
    }

    // Accessibility checks
    if (!method.isPublic() && !isAccessible()) {
      VM_Class accessingClass = VM_Class.getClassFromStackFrame(1);
      JikesRVMSupport.checkAccess(method, accessingClass);
    }

    // find the right method to call
    if (! method.isStatic()) {
        VM_Class C = VM_Magic.getObjectType(receiver).asClass(); 
        method = C.findVirtualMethod(method.getName(), method.getDescriptor());
    }
    
    // Forces initialization of declaring class
    if (method.isStatic() && !declaringClass.isInitialized()) {
      try {
        VM_Runtime.initializeClassForDynamicLink(declaringClass);
      } catch (Throwable e) {
        ExceptionInInitializerError ex = new ExceptionInInitializerError();
        ex.initCause(e);
        throw ex;
      }
    }

    // Invoke method
    try {
      return VM_Reflection.invoke(method, receiver, args);
    } catch (Throwable t) {
      throw new InvocationTargetException(t,
                "While invoking the method:\n\t\"" + method + "\"\n"
                + " on the object:\n\t\"" + receiver + "\"\n"
                + " it threw the exception:\n\t\"" + t + "\"");
     }
  }

  public String toString() {
    Class current;
        
    StringBuffer buf = new StringBuffer();
    String mods = Modifier.toString(getModifiers());
    if(mods.length() != 0) {
      buf.append(mods);
      buf.append(" ");
    }
        
    current = getReturnType();
    int arity = 0;
    while(current.isArray()) {
      current = current.getComponentType();
      arity++;
    }
    buf.append(current.getName());
    for(;arity > 0; arity--) buf.append("[]");
        
    buf.append(" ");
    buf.append(getDeclaringClass().getName());
    buf.append(".");
    buf.append(getName());
    buf.append("(");
    Class[] types = getParameterTypes();
    for(int i = 0; i < types.length; i++) {
      current = types[i];
      arity = 0;
      while(current.isArray()) {
        current = current.getComponentType();
        arity++;
      }
      buf.append(current.getName());
      for(;arity > 0; arity--) buf.append("[]");
      if(i != (types.length - 1))
        buf.append(",");
    }
    buf.append(")");
    types = getExceptionTypes();
    if(types.length > 0) {
      buf.append(" throws ");
      for(int i = 0; i < types.length; i++) {
        current = types[i];
        buf.append(current.getName());
        if(i != (types.length - 1))
          buf.append(",");
      }
    }

    return buf.toString();
  }
}
