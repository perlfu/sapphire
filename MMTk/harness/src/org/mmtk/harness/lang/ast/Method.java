/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package org.mmtk.harness.lang.ast;

import java.util.List;

import org.mmtk.harness.lang.Visitor;
import org.mmtk.harness.lang.parser.Token;

/**
 * Abstract superclass of methods implemented in a variety of ways
 */
public abstract class Method extends AbstractAST implements Statement, Expression, Comparable<Method> {

  /** The name of this block */
  protected final String name;

  /** Number of parameters */
  protected final int params;

  protected final Type returnType;

  /**
   * Constructor.
   * @param name
   * @param params
   */
  protected Method(Token t, String name, int params, Type returnType) {
    super(t);
    this.name = name;
    this.params = params;
    this.returnType = returnType;
  }

  protected Method(String name, int params, Type returnType) {
    super(0,0);
    this.name = name;
    this.params = params;
    this.returnType = returnType;
  }

  /**
   * Get the name of this method.
   */
  public String getName() {
    return name;
  }

  public void accept(Visitor v) {
    v.visit(this);
  }

  public int getParamCount() {
    return params;
  }

  public abstract List<Type> getParamTypes();

  public Type getReturnType() { return returnType; }

  public Method getMethod() {
    return this;
  }

  @Override
  public int compareTo(Method m) {
    return name.compareTo(m.getName());
  }

  @Override
  public boolean equals(Object o) {
    if (o instanceof Method) {
      return compareTo((Method)o) == 0;
    } else {
      return false;
    }
  }

  @Override
  public int hashCode() {
    return name.hashCode();
  }
}
