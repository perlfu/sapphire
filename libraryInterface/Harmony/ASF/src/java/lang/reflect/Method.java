/* 
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package java.lang.reflect;

import java.lang.annotation.Annotation;
import java.lang.reflect.VMCommonLibrarySupport;

import org.jikesrvm.classloader.RVMClass;
import org.jikesrvm.classloader.RVMMember;
import org.jikesrvm.classloader.RVMMethod;
import org.jikesrvm.runtime.Reflection;
import org.jikesrvm.runtime.ReflectionBase;

/**
 * This class must be implemented by the VM vendor. This class models a method.
 * Information about the method can be accessed, and the method can be invoked
 * dynamically.
 * 
 */
public final class Method extends AccessibleObject implements GenericDeclaration, Member {
    private final RVMMethod vmMethod;
    private final ReflectionBase invoker;

    /**
     * Constructor
     */
    Method(RVMMethod vmMethod){
      this.vmMethod = vmMethod;
      if (Reflection.cacheInvokerInJavaLangReflect) {
        invoker = vmMethod.getInvoker();
      } else {
        invoker = null;
      }
    }
    
    /**
     * Prevent this class from being instantiated
     */
    private Method(){
      vmMethod = null;
      invoker = null;
    }
    
    public TypeVariable<Method>[] getTypeParameters() {
      throw new Error("TODO");
    }

    /**
     * <p>
     * Returns the String representation of the method's declaration, including
     * the type parameters.
     * </p>
     * 
     * @return An instance of String.
     * @since 1.5
     */
    public String toGenericString() {
      throw new Error("TODO");
    }

    /**
     * <p>
     * Gets the parameter types as an array of {@link Type} instances, in
     * declaration order. If the method has no parameters, then an empty array
     * is returned.
     * </p>
     * 
     * @return An array of {@link Type} instances.
     * @throws GenericSignatureFormatError if the generic method signature is
     *         invalid.
     * @throws TypeNotPresentException if the component type points to a missing
     *         type.
     * @throws MalformedParameterizedTypeException if the component type points
     *         to a type that can't be instantiated for some reason.
     * @since 1.5
     */
    public Type[] getGenericParameterTypes() {
      throw new Error("TODO");
    }

    /**
     * <p>
     * Gets the exception types as an array of {@link Type} instances. If the
     * method has no declared exceptions, then an empty array is returned.
     * </p>
     * 
     * @return An array of {@link Type} instances.
     * @throws GenericSignatureFormatError if the generic method signature is
     *         invalid.
     * @throws TypeNotPresentException if the component type points to a missing
     *         type.
     * @throws MalformedParameterizedTypeException if the component type points
     *         to a type that can't be instantiated for some reason.
     * @since 1.5
     */
    public Type[] getGenericExceptionTypes() {
      throw new Error("TODO");
    }

    /**
     * <p>
     * Gets the return type as a {@link Type} instance.
     * </p>
     * 
     * @return A {@link Type} instance.
     * @throws GenericSignatureFormatError if the generic method signature is
     *         invalid.
     * @throws TypeNotPresentException if the component type points to a missing
     *         type.
     * @throws MalformedParameterizedTypeException if the component type points
     *         to a type that can't be instantiated for some reason.
     * @since 1.5
     */
    public Type getGenericReturnType() {
      throw new Error("TODO");
    }

    /**
     * <p>
     * Gets an array of arrays that represent the annotations of the formal
     * parameters of this method. If there are no parameters on this method,
     * then an empty array is returned. If there are no annotations set, then
     * and array of empty arrays is returned.
     * </p>
     * 
     * @return An array of arrays of {@link Annotation} instances.
     * @since 1.5
     */
    public Annotation[][] getParameterAnnotations() {
      return super.getParameterAnnotations();
    }

    /**
     * <p>
     * Indicates whether or not this method takes a variable number argument.
     * </p>
     * 
     * @return A value of <code>true</code> if a vararg is declare, otherwise
     *         <code>false</code>.
     * @since 1.5
     */
    public boolean isVarArgs() {
      return super.isVarArgs();
    }

    /**
     * <p>
     * Indicates whether or not this method is a bridge.
     * </p>
     * 
     * @return A value of <code>true</code> if this method's a bridge,
     *         otherwise <code>false</code>.
     * @since 1.5
     */
    public boolean isBridge() {
		return (vmMethod.getModifiers() & Modifier.BRIDGE) != 0;
    }

    public boolean isSynthetic() {
		return super.isSynthetic();
    }
    
    /**
     * <p>Gets the default value for the annotation member represented by
     * this method.</p>
     * @return The default value or <code>null</code> if none.
     * @throws TypeNotPresentException if the annotation is of type {@link Class}
     * and no definition can be found.
     * @since 1.5
     */
    public Object getDefaultValue() {
        return vmMethod.getAnnotationDefault();
    }
    
	/**
	 * Compares the specified object to this Method and answer if they are
	 * equal. The object must be an instance of Method with the same defining
	 * class and parameter types.
	 * 
	 * @param object
	 *            the object to compare
	 * @return true if the specified object is equal to this Method, false
	 *         otherwise
	 * @see #hashCode
	 */
	@Override
   public boolean equals(Object object) {
	  if (object instanceof Method) {
		 Method that = (Method)object;
		 return this.vmMethod == that.vmMethod;
	  } else {
		 return false;
	  }
	}

	/**
	 * Return the {@link Class} associated with the class that defined this
	 * method.
	 * 
	 * @return the declaring class
	 */
	public Class<?> getDeclaringClass() {
      return (Class<?>)vmMethod.getDeclaringClass().getClassForType();
	}

	/**
	 * Return an array of the {@link Class} objects associated with the
	 * exceptions declared to be thrown by this method. If the method was not
	 * declared to throw any exceptions, the array returned will be empty.
	 * 
	 * @return the declared exception classes
	 */
	public Class<?>[] getExceptionTypes() {
		return super.getExceptionTypes();
	}

	/**
	 * Return the modifiers for the modeled method. The Modifier class
	 * should be used to decode the result.
	 * 
	 * @return the modifiers
	 * @see java.lang.reflect.Modifier
	 */
    public int getModifiers() {
		return super.getModifiers();
	}

	/**
	 * Return the name of the modeled method.
	 * 
	 * @return the name
	 */
	public String getName() {
		return vmMethod.getName().toString();
	}

	/**
	 * Return an array of the {@link Class} objects associated with the
	 * parameter types of this method. If the method was declared with no
	 * parameters, the array returned will be empty.
	 * 
	 * @return the parameter types
	 */
	public Class<?>[] getParameterTypes() {
      return VMCommonLibrarySupport.typesToClasses(vmMethod.getParameterTypes());
	}

	/**
	 * Return the {@link Class} associated with the return type of this
	 * method.
	 * 
	 * @return the return type
	 */
	public Class<?> getReturnType() {
	  return vmMethod.getReturnType().resolve().getClassForType();
	}

	/**
	 * Answers an integer hash code for the receiver. Objects which are equal
	 * answer the same value for this method. The hash code for a Method is the
	 * hash code of the method's name.
	 * 
	 * @return the receiver's hash
	 * @see #equals
	 */
	@Override
    public int hashCode() {
		return getName().hashCode();
	}

	/**
	 * Return the result of dynamically invoking the modeled method. This
	 * reproduces the effect of
	 * <code>receiver.methodName(arg1, arg2, ... , argN)</code> This method
	 * performs the following:
	 * <ul>
	 * <li>If the modeled method is static, the receiver argument is ignored.
	 * </li>
	 * <li>Otherwise, if the receiver is null, a NullPointerException is
	 * thrown.</li>
	 * If the receiver is not an instance of the declaring class of the method,
	 * an IllegalArgumentException is thrown.
	 * <li>If this Method object is enforcing access control (see
	 * AccessibleObject) and the modeled method is not accessible from the
	 * current context, an IllegalAccessException is thrown.</li>
	 * <li>If the number of arguments passed and the number of parameters do
	 * not match, an IllegalArgumentException is thrown.</li>
	 * <li>For each argument passed:
	 * <ul>
	 * <li>If the corresponding parameter type is a base type, the argument is
	 * unwrapped. If the unwrapping fails, an IllegalArgumentException is
	 * thrown.</li>
	 * <li>If the resulting argument cannot be converted to the parameter type
	 * via a widening conversion, an IllegalArgumentException is thrown.</li>
	 * </ul>
	 * <li>If the modeled method is static, it is invoked directly. If it is
	 * non-static, the modeled method and the receiver are then used to perform
	 * a standard dynamic method lookup. The resulting method is then invoked.
	 * </li>
	 * <li>If an exception is thrown during the invocation it is caught and
	 * wrapped in an InvocationTargetException. This exception is then thrown.
	 * </li>
	 * <li>If the invocation completes normally, the return value is itself
	 * returned. If the method is declared to return a base type, the return
	 * value is first wrapped. If the return type is void, null is returned.
	 * </li>
	 * </ul>
	 * 
	 * @param receiver
	 * 	          The object on which to call the modeled method
	 * @param args
	 *            the arguments to the method
	 * @return the new, initialized, object
	 * @throws java.lang.NullPointerException
	 *                if the receiver is null for a non-static method
	 * @throws java.lang.IllegalAccessException
	 *                if the modeled method is not accessible
	 * @throws java.lang.IllegalArgumentException
	 *                if an incorrect number of arguments are passed, the
	 *                receiver is incompatible with the declaring class, or an
	 *                argument could not be converted by a widening conversion
	 * @throws java.lang.reflect.InvocationTargetException
	 *                if an exception was thrown by the invoked method
	 * @see java.lang.reflect.AccessibleObject
	 */
	public Object invoke(Object receiver, Object... args)
			throws IllegalAccessException, IllegalArgumentException,
			InvocationTargetException {
	    return VMCommonLibrarySupport.invoke(receiver, args, vmMethod, this, RVMClass.getClassFromStackFrame(1), invoker);
	}

	/**
	 * Answers a string containing a concise, human-readable description of the
	 * receiver. The format of the string is modifiers (if any) return type
	 * declaring class name '.' method name '(' parameter types, separated by
	 * ',' ')' If the method throws exceptions, ' throws ' exception types,
	 * separated by ',' For example:
	 * <code>public native Object java.lang.Method.invoke(Object,Object) throws IllegalAccessException,IllegalArgumentException,InvocationTargetException</code>
	 * 
	 * @return a printable representation for the receiver
	 */
	@Override
   public String toString() {
	  StringBuilder sb = new StringBuilder(80);
	  // append modifiers if any
	  int modifier = getModifiers();
	  if (modifier != 0) {
		 // BRIDGE & VARARGS recognized incorrectly
		 final int MASK = ~(Modifier.BRIDGE | Modifier.VARARGS);
		 sb.append(Modifier.toString(modifier & MASK)).append(' ');            
	  }
	  // append return type
	  appendArrayType(sb, getReturnType());
	  sb.append(' ');
	  // append full method name
	  sb.append(getDeclaringClass().getName()).append('.').append(getName());
	  // append parameters
	  sb.append('(');
	  appendArrayType(sb, getParameterTypes());
	  sb.append(')');
	  // append exeptions if any
	  Class[] exn = getExceptionTypes(); 
	  if (exn.length > 0) {
		 sb.append(" throws ");
		 appendSimpleType(sb, exn);
	  }
	  return sb.toString();
	}

  /* ---- Non-API methods ---- */

  /**
   * Get the VM member implementation. Package protected to stop outside use.
   */
  @Override
  RVMMember getVMMember() {
	 return vmMethod;
  }

  /**
   * Get the VM method implementation, invalid for fields. Package protected to
   * stop outside use.
   */
  @Override
  RVMMethod getVMMethod() {
	 return vmMethod;
  }
}
