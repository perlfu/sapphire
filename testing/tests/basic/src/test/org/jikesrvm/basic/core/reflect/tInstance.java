/*
 * This file is part of Jikes RVM (http://jikesrvm.sourceforge.net).
 * The Jikes RVM project is distributed under the Common Public License (CPL).
 * A copy of the license is included in the distribution, and is also
 * available at http://www.opensource.org/licenses/cpl1.0.php
 *
 * (C) Copyright IBM Corp. 2001
 */
package test.org.jikesrvm.basic.core.reflect;

import java.lang.reflect.Method;

/**
 * @author unascribed
 */
class tInstance {
  public int ifield;
  public double dfield;
  public boolean bfield;
  public Object ofield;

  public static void gc() {
    System.gc();
  }

  public int iuserFunction(int i) {
    gc();
    return ifield + i;
  }

  public double duserFunction(double d) {
    gc();
    return d + dfield;
  }

  public boolean buserFunction(boolean x) {
    gc();
    if (x)
      return bfield;
    else
      return x;
  }

  public Object ouserFunction(String s) {
    gc();
    ofield = s + "abc";
    return ofield;
  }

  public void vuserFunction(int i, Integer x) {
    gc();
    ifield = ifield + i + x;
  }

  public static void main(String[] args) throws Exception {
    // Class.forName
    //
    Class c = Class.forName("test.org.jikesrvm.basic.core.reflect.tInstance");
    tInstance myInstance = (tInstance) c.newInstance();
    myInstance.ifield = 4;
    myInstance.dfield = 8.8;
    myInstance.bfield = true;
    myInstance.ofield = null;

    System.out.println(c);

    // Class.getMethods  Method.getName
    //
    Method[] methods = c.getMethods();
    Method imethod = null;
    Method dmethod = null;
    Method bmethod = null;
    Method omethod = null;
    Method vmethod = null;

    for (Method method : methods) {
      String methodName = method.getName();
      if (methodName.equals("iuserFunction"))
        imethod = method;
      else if (methodName.equals("duserFunction"))
        dmethod = method;
      else if (methodName.equals("buserFunction"))
        bmethod = method;
      else if (methodName.equals("ouserFunction"))
        omethod = method;
      else if (methodName.equals("vuserFunction"))
        vmethod = method;
    }

    Object[] methodargs = new Object[1];

    /******************/
    //  int           //
    /******************/

    if (imethod == null) {
      System.out.println("tInstance.iuserFunction not found!");
      System.exit(1);
    } else {
      System.out.println("================= READY TO CALL: " + imethod);
      methodargs[0] = 3;
      int iresult = (Integer) imethod.invoke(myInstance, methodargs);
      if (iresult != 7) {
        System.out.println("Wrong answer from iuserFunction");
        System.out.println(iresult);
        System.exit(1);
      }
    }

    /******************/
    //  double        //
    /******************/

    if (dmethod == null) {
      System.out.println("tInstance.duserFunction not found!");
      System.exit(1);
    } else {
      System.out.println("================= READY TO CALL: " + dmethod);
      methodargs[0] = 3.4;
      double dresult = (Double) dmethod.invoke(myInstance, methodargs);
      if (dresult < 12.2 || dresult >= 12.2000001) {
        System.out.println("Wrong answer from duserFunction");
        System.out.println(dresult);
        System.exit(1);
      }
    }

    /******************/
    //  boolean       //
    /******************/

    if (bmethod == null) {
      System.out.println("tInstance.buserFunction not found!");
      System.exit(1);
    } else {
      System.out.println("================= READY TO CALL: " + bmethod);
      methodargs[0] = Boolean.TRUE;
      boolean bresult = (Boolean) bmethod.invoke(myInstance, methodargs);
      if (bresult != true) {
        System.out.println("Wrong answer from buserFunction");
        System.exit(1);
      }
    }

    /******************/
    //  Object        //
    /******************/

    if (omethod == null) {
      System.out.println("tInstance.ouserFunction not found!");
      System.exit(1);
    } else {
      System.out.println("================= READY TO CALL: " + omethod);
      methodargs[0] = "123";
      Object oresult = omethod.invoke(myInstance, methodargs);
      if (!(oresult instanceof java.lang.String) || !oresult.equals("123abc")) {
        System.out.println("Wrong answer from ouserFunction");
        System.exit(1);
      }
    }

    /*******************/
    //  Void  two args //
    /*******************/

    if (vmethod == null) {
      System.out.println("tInstance.vuserFunction not found!");
      System.exit(1);
    } else {
      System.out.println("================= READY TO CALL: " + vmethod);
      Object[] twoargs = new Object[2];
      twoargs[0] = 4;
      twoargs[1] = 10;
      Object vresult = vmethod.invoke(myInstance, twoargs);
      if ((vresult != null) || (myInstance.ifield != 18)) {
        System.out.println("Wrong results from vuserFunction");
        System.exit(1);
      }
    }
    System.out.println("Test success");

  }
}
