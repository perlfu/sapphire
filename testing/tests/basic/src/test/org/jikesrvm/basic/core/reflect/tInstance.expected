class test.org.jikesrvm.basic.core.reflect.tInstance
================= READY TO CALL: public int test.org.jikesrvm.basic.core.reflect.tInstance.iuserFunction(int)
================= READY TO CALL: public double test.org.jikesrvm.basic.core.reflect.tInstance.duserFunction(double)
================= READY TO CALL: public boolean test.org.jikesrvm.basic.core.reflect.tInstance.buserFunction(boolean)
================= READY TO CALL: public java.lang.Object test.org.jikesrvm.basic.core.reflect.tInstance.ouserFunction(java.lang.String)
================= READY TO CALL: public void test.org.jikesrvm.basic.core.reflect.tInstance.vuserFunction(int,java.lang.Integer)
Test success
